### PR TITLE
add hook for disabling the default extensions in the build output

### LINKED ISSUE #10690: Add  hook for excluding certain build output extensions from the nuget package
Currently there is no way to exclude certain output extensions from the package.

https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets#L48

adding something like  should do the trick

```
<AllowedOutputExtensionsInPackageBuildOutputFolder Condition="'$(AllowedOutputExtensionsInPackageBuildOutputFolder)' == ''">.dll; .exe; .winmd; .json; .pri; .xml; $(AllowedOutputExtensionsInPackageBuildOutputFolder)</AllowedOutputExtensionsInPackageBuildOutputFolder>
```

 cc @safern @nkolev92 @ericstj @viktorhofer

### PR DESCRIPTION
**Pull request by Anipik**
Fixes: https://github.com/NuGet/Home/issues/10690

## Description

This msbuild flag allows user to edit the default file extensions included in the build output of the package. eg. it allows us to exclude the .xml files from the package. 

Test are added for validation

cc @ViktorHofer @ericstj

### CODE REVIEW

File: src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets [hunk 1]
```diff
@@ -45,8 +45,8 @@ Copyright (c) .NET Foundation. All rights reserved.
     <AddPriFileDependsOn Condition="'$(MicrosoftPortableCurrentVersionPropsHasBeenImported)' == 'true'">DeterminePortableBuildCapabilities</AddPriFileDependsOn>
     <WarnOnPackingNonPackableProject Condition="'$(WarnOnPackingNonPackableProject)' == ''">false</WarnOnPackingNonPackableProject>
     <ImportNuGetBuildTasksPackTargetsFromSdk Condition="'$(ImportNuGetBuildTasksPackTargetsFromSdk)' == ''">false</ImportNuGetBuildTasksPackTargetsFromSdk>
-    <AllowedOutputExtensionsInPackageBuildOutputFolder>.dll; .exe; .winmd; .json; .pri; .xml; $(AllowedOutputExtensionsInPackageBuildOutputFolder)</AllowedOutputExtensionsInPackageBuildOutputFolder>
-    <AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder Condition="'$(SymbolPackageFormat)' != 'snupkg'">.pdb; .mdb; $(AllowedOutputExtensionsInPackageBuildOutputFolder); $(AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder)</AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder>
+    <DefaultAllowedOutputExtensionsInPackageBuildOutputFolder Condition="'$(DefaultAllowedOutputExtensionsInPackageBuildOutputFolderCondition)' == ''">dll; .exe; .winmd; .json; .pri; .xml</DefaultAllowedOutputExtensionsInPackageBuildOutputFolder>
```
Comment (Nirmal4G) [inline src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets]:
```suggestion
    <DefaultAllowedOutputExtensionsInPackageBuildOutputFolder Condition="'$(EnableDefaultOutputExtensionsInPackageBuildOutputFolder)' != 'false'">dll; .exe; .winmd; .json; .pri; .xml</DefaultAllowedOutputExtensionsInPackageBuildOutputFolder>
```

OR


```suggestion
    <EnableDefaultOutputExtensionsInPackageBuildOutputFolder Condition="'$(EnableDefaultOutputExtensionsInPackageBuildOutputFolder)' == ''">true</EnableDefaultOutputExtensionsInPackageBuildOutputFolder>
    <DefaultAllowedOutputExtensionsInPackageBuildOutputFolder Condition="'$(EnableDefaultOutputExtensionsInPackageBuildOutputFolder)' == 'true'">dll; .exe; .winmd; .json; .pri; .xml</DefaultAllowedOutputExtensionsInPackageBuildOutputFolder>
```

Similar to what .NET SDK targets use.

**Review by Nirmal4G — COMMENTED** [review]
Property Name suggestion...

File: src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets (hunk 1 above)
Comment (Anipik) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets]: Introducing an extra ```EnableDefaultOutputExtensionsInPackageBuildOutputFolder``` wont give us any extra use and users will have to set 2 props to change the default extensions

Comment (nkolev92) [inline src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets]: you probably don't need the conditional set of the default prop. 

It's the latter prop that matters.

Comment (Anipik) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets]: can you elaborate ?

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets]: I think my response to another comment that seems to be gone now. 

I see you are using `DefaultAllowedOutputExtensionsInPackageBuildOutputFolder` and `DefaultAllowedOutputExtensionsInPackageBuildOutputFolderCondition`. Do you need the 2nd one?

```xml
<DefaultA Condition=" '$(DefaultA)' == '' ">defaults</DefaultA>
<A>$(DefaultA);$(A)</A>

Comment (Anipik) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets]: ah right, i thought i removed the "condition" substring. i pushed the commit to correct it.

File: test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs [hunk 2]
```diff
@@ -3810,6 +3810,45 @@ public void PackCommand_BuildOutput_DoesNotContainForSpecificFramework(string fr
...
+
+                var nupkgPath = Path.Combine(workingDirectory, $"{projectName}.1.0.0.nupkg");
+                var nuspecPath = Path.Combine(workingDirectory, "obj", $"{projectName}.1.0.0.nuspec");
+                Assert.True(File.Exists(nupkgPath), "The output .nupkg is not in the expected place");
+                Assert.True(File.Exists(nuspecPath), "The intermediate nuspec file is not in the expected place");
+
+                // Assert
+                using (var nupkgReader = new PackageArchiveReader(nupkgPath))
+                {
+                    var buildItems = nupkgReader.GetBuildItems().ToList();
```
Comment (nkolev92) [inline test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs]: Does this pass?

This sohuld only capture build files?

I'd recommend looking at the exact file list, as then you are not at the mercy of the implementation of the GetBuildItems method.

Comment (Anipik) [reply test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs]: i tested the fix standalone.
You are right, this should fail. i was not able to run the test using the local copy of  targets file, it end up using the my machine copy ? how can i fix that and run the test

Comment (nkolev92) [reply test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs]: The tests should be copying the internal targets.

You can debug it here: https://github.com/NuGet/NuGet.Client/blob/7e3b0e445edecbdb6c46895ee4375cf834f45d87/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/MsbuildIntegrationTestFixture.cs#L570-L571

**Comment by nkolev92** [comment]
🔔 @Anipik

**Comment by Anipik** [comment]
i will try to get the tests in next week.

File: test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs (hunk 2 above)
Comment (Anipik) [reply test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs]: > C:\git\NuGet.Client\.test\work\cd78165c\b5afa28a\sdk\5.0.200-servicing.21120.4\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(279,15): error MSB4064: The "Readme" parameter is not supported by the "PackTask" task. Verify the parameter exists on the task, and it is a settable public instance property. [C:\git\NuGet.Client\.test\work\c783b310\bbfe3ac5\ClassLibrary1\ClassLibrary1.csproj]
C:\git\NuGet.Client\.test\work\cd78165c\b5afa28a\sdk\5.0.200-servicing.21120.4\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(222,5): error MSB4063: The "PackTask" task could not be initialized with its input parameters.  [C:\git\NuGet.Client\.test\work\c783b310\bbfe3ac5\ClassLibrary1\ClassLibrary1.csproj]

i was able to run the tests but i get this as an error. i verified the file being copied is correct. @nkolev92 can u help me with this ?

Comment (nkolev92) [reply test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs]: > The "Readme" parameter is not supported by the "PackTask" task

hmm, this suggests to me that targets and dlls are somehow out of sync. 

Maybe try clearing the artifacts folder, build from the commandline by running `msbuild /restore NuGet.sln /v:m`, and then running the test again. 
If not of that works, try pinging me offline.

Comment (Anipik) [reply test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs]: @nkolev92 that worked thanks. I modified to check for the specific file and test fails before and passes after the change.

**Comment by Anipik** [comment]
@nkolev92 it seems like the ci is hung on this, do we need to do something ?

**Comment by nkolev92** [comment]
@anipik the CI for people without write access doesn't run automatically.

Kicked it off.

**Comment by nkolev92** [comment]
@Anipik 

There's some test failures. Can you access the [following](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4778032&view=ms.vss-test-web.build-test-results-tab&runId=21839872&resultId=100106&paneView=debug), check out the test table for partial results. 

The tests that are failing are pack specific.

File: src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets [hunk 3]
```diff
@@ -45,7 +45,8 @@ Copyright (c) .NET Foundation. All rights reserved.
     <AddPriFileDependsOn Condition="'$(MicrosoftPortableCurrentVersionPropsHasBeenImported)' == 'true'">DeterminePortableBuildCapabilities</AddPriFileDependsOn>
     <WarnOnPackingNonPackableProject Condition="'$(WarnOnPackingNonPackableProject)' == ''">false</WarnOnPackingNonPackableProject>
     <ImportNuGetBuildTasksPackTargetsFromSdk Condition="'$(ImportNuGetBuildTasksPackTargetsFromSdk)' == ''">false</ImportNuGetBuildTasksPackTargetsFromSdk>
-    <AllowedOutputExtensionsInPackageBuildOutputFolder>.dll; .exe; .winmd; .json; .pri; .xml; $(AllowedOutputExtensionsInPackageBuildOutputFolder)</AllowedOutputExtensionsInPackageBuildOutputFolder>
+    <DefaultAllowedOutputExtensionsInPackageBuildOutputFolder Condition="'$(DefaultAllowedOutputExtensionsInPackageBuildOutputFolder)' == ''">dll; .exe; .winmd; .json; .pri; .xml</DefaultAllowedOutputExtensionsInPackageBuildOutputFolder>
```
Comment (erdembayar) [inline src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets]: You're missing `.` in front of dll

File: test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs [hunk 4]
```diff
@@ -3810,6 +3810,45 @@ public void PackCommand_BuildOutput_DoesNotContainForSpecificFramework(string fr
...
+                var nupkgPath = Path.Combine(workingDirectory, $"{projectName}.1.0.0.nupkg");
+                var nuspecPath = Path.Combine(workingDirectory, "obj", $"{projectName}.1.0.0.nuspec");
+                Assert.True(File.Exists(nupkgPath), "The output .nupkg is not in the expected place");
+                Assert.True(File.Exists(nuspecPath), "The intermediate nuspec file is not in the expected place");
+
+                // Assert
+                using (var nupkgReader = new PackageArchiveReader(nupkgPath))
+                {
+                    var allFiles = nupkgReader.GetFiles().ToList();
+                    Assert.DoesNotContain($"lib/net5.0/{projectName}.xml", allFiles);
```
Comment (erdembayar) [inline test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:3848]: I feel we're only testing one side which is not include `xml` file. Could be able to assert if files with only with `.dll` extension are in package which is more resemble what're trying to achieve?

Comment (Anipik) [reply test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:3848]: i actually had that earlier but found out that some of the files are always include in the package. as rel.rels, nuspec

**Comment by Anipik** [comment]
@nkolev92 do we need to trigger the ci again here ?

File: test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs (hunk 4 above)
Comment (erdembayar) [reply test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:3848]: Another way is add assert that package doesn't include `.dll; .exe; .winmd; .json; .pri` files.
How about that? Then I'll re-trigger the ci for you?

Comment (Anipik) [reply test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:3847]: > Another way is add assert that package doesn't include .dll; .exe; .winmd; .json; .pri files.
How about that? Then I'll re-trigger the ci for you?

u mean every extension except the .dll ?

Comment (Anipik) [reply test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:3848]: i added the other doesnot contain tests. can you retrigger the ci

Comment (erdembayar) [reply test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:3848]: Can we still have `Assert.Contain($"lib/net5.0/{projectName}.dll", allFiles);`?

Comment (erdembayar) [reply test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:3848]: This one could be less fragile and more covering. `Assert.False(allFiles.Any(f => f.EndsWith(".exe")));`

Comment (Anipik) [reply test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:3848]: done, i kept the xml to actually test the removal

Comment (erdembayar) [reply test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:3848]: How about doing same for other types too? Most likely that is all.
Assert.False(allFiles.Any(f => f.EndsWith(".winmd")));
Assert.False(allFiles.Any(f => f.EndsWith(".pri")));
Assert.False(allFiles.Any(f => f.EndsWith(".json")));

**Comment by erdembayar** [comment]
@Anipik 
I just trigged CI build. 🤞

**Comment by Anipik** [comment]
@erdembayar @nkolev92 the ci passed, can you merge it as well ?

**Review by erdembayar — APPROVED** [review]
I'll approve it now, but wait little bit before merge, let other people review too.

**Review by nkolev92 — APPROVED** [review]
LGTM.

**Comment by Anipik** [comment]
hey @erdembayar @nkolev92 do we need something more here ?

**Comment by erdembayar** [comment]
> hey @erdembayar @nkolev92 do we need something more here ?

@Anipik 
Don't worry, I'll merge by lunch time if there is no one else comment.

**Comment by Anipik** [comment]
awesome thanks.

**Pull request merged**